File: ass1/ass_1_2.py
R = 20000000
N = 500
K = 200
B = R / N

dp = [['*' for _ in range(K + 1)] for _ in range(6)]

def calculate(dp, index, k):
    if index == 1:
        dp[index][k] = k * B - 2 * dp[2][k] - 3 * dp[3][k] - 4 * dp[4][k] - 5 * dp[5][k]
        return
    dp[index][k] = ((6 - index) * dp[index - 1][k - 1]) / (N - k + 1) \
                 + dp[index][k - 1] \
                 - ((6 - index - 1) * dp[index][k - 1]) / (N - k + 1)

dp[1][1] = B;
for i in range(2, 6):
    dp[i][1] = 0
# for i in dp:
#     print(i)

for i in range(2, K + 1):
    for j in [5, 4, 3, 2, 1]:
        calculate(dp, j, i)

for i in dp:
    print(i)



# R = 20000000
# N = 4000
# K = 3
# B = 750

# dp = [[0 for _ in range(K + 1)] for _ in range(4)]

# def calculate(dp, index, k):
#     if index == 1:
#         dp[index][k] = k * B - 2 * dp[2][k] - 3 * dp[3][k]
#         return
#     dp[index][k] = (4 - index) * dp[index - 1][k - 1] / (N - k + 1) \
#                  + dp[index][k - 1] \
#                  - (3 - index) * dp[index][k - 1] / (N - k - 1)

# dp[1][1] = B;
# for i in range(2, K + 1):
#     for j in [3, 2, 1]:
#         calculate(dp, j, i)

# for i in dp:
#     print(i)

File: ass1/ass_1_2.cpp
#include<iostream>
#include<vector>

using namespace std;

void calculate(vector<vector<double>> & dp, int index, int k, long double R, long double N, long double B){
    if(index == 1){
        dp[index][k] = k * B - 2 * dp[2][k] - 3 * dp[3][k] - 4 * dp[4][k] - 5 * dp[5][k];
        return;
    }
    dp[index][k] = (6 - index) * dp[index - 1][k - 1] / (N - k + 1) 
                 + dp[index][k - 1] 
                 - (5 - index) * dp[index][k - 1] / (N - k + 1);
}

int main(){
    long double R = 20000000;
    long double N = 500;
    int K = 200;
    long double B = R / N;

    vector<vector<double>> dp(5 + 1, vector<double>(K + 1));
    dp[1][1] = B;

    for(int i = 2; i <= K; ++i){
        for(int j = 5; j > 0; --j){
            calculate(dp, j, i, R, N, B);
        }
    }
    // for(int i = 1; i < 6; ++i){
    //     for(int j = 1; j < 201; ++j){
    //         cout << dp[i][j] << " ";
    //     }
    //     putchar(10);
    // }
    cout << dp[5][200] << endl;
}